pkg/networkpolicy/packet.go

```diff
@@ -29,59 +33,111 @@ func (p packet) String() string {
 		t.dstIP = make(net.IP, net.IPv6len)
 		copy(t.dstIP, b[24:40])
-		// NextHeader (not extension headers supported)
-		protocol = int(b[6])
+		// Handle extension headers.
+		nxtHeader = int(b[6])
+		l4offset = 40
+		for nxtHeader == syscall.IPPROTO_DSTOPTS || nxtHeader == syscall.IPPROTO_HOPOPTS || nxtHeader == syscall.IPPROTO_ROUTING {
+			// These headers have a lenght in 8-octet units, not
+			// including the first 8 octets
+			nxtHeader = int(b[l4offset])
+			l4offset += (8 + int(b[l4offset+1])*8)
+			// Now l4offset points to either another extension header,
+			// or an L4 header. So we must have at least 8 byte data
+			// after this (minimum extension header size)
+			if (l4offset + 8) >= len(b) {
+				return t, ErrorTooShort
+			}
+		}
+		if nxtHeader == syscall.IPPROTO_FRAGMENT {
+			// Only the first fragment has the L4 header
+			fragOffset := int(binary.BigEndian.Uint16(b[l4offset+2 : l4offset+4]))
+			if fragOffset&0xfff8 == 0 {
+				nxtHeader = int(b[l4offset])
+				l4offset += 8
+				// Here it's assumed that the fragment is the last
+				// extension header before the L4 header. But more
+				// IPPROTO_DSTOPTS are allowed by the recommended order.
+				// TODO: handle extra IPPROTO_DSTOPTS.
+			} else {
+				// If this is NOT the first fragment, we have no L4
+				// header and the payload begins after this
+				// header. Return a packet with t.proto unset
+				return t, nil
+			}
+		}
+		protocol = nxtHeader
 	default:
-		return t, fmt.Errorf("unknown versions %d", version)
+		return t, fmt.Errorf("unknown version %d", version)
 	}
 
-	var dataOffset int
 	switch protocol {
-	case 6:
+	case syscall.IPPROTO_TCP:
 		t.proto = v1.ProtocolTCP
-		dataOffset = int(b[hdrlen+12] >> 4) // data offset
-	case 17:
+		dataOffset := int(b[l4offset+12]>>4) * 4
+		if dataOffset < 20 {
+			return t, ErrorCorrupted
+		}
+		payloadOffset = l4offset + dataOffset
+	case syscall.IPPROTO_UDP:
 		t.proto = v1.ProtocolUDP
-		dataOffset = hdrlen + 8 // data starts after
-	case 132:
+		payloadOffset = l4offset + 8
+	case syscall.IPPROTO_SCTP:
 		t.proto = v1.ProtocolSCTP
-		dataOffset = hdrlen + 8
+		payloadOffset = l4offset + 8
 	default:
-		return t, fmt.Errorf("unknown protocol %d", protocol)
+		// Return a packet with t.proto unset, and ports 0
+		return t, nil
+
 	}
-	// TCP, UDP and SCTP srcPort and dstPort are the first 4 bytes after the IP header
-	t.srcPort = int(binary.BigEndian.Uint16(b[hdrlen : hdrlen+2]))
-	t.dstPort = int(binary.BigEndian.Uint16(b[hdrlen+2 : hdrlen+4]))
+	if payloadOffset > len(b) {
+		// We must have a complete L4 header, however payload may be zero-lengh
+		return t, ErrorTooShort
+	}
+	t.srcPort = int(binary.BigEndian.Uint16(b[l4offset : l4offset+2]))
+	t.dstPort = int(binary.BigEndian.Uint16(b[l4offset+2 : l4offset+4]))
+
 	// Obtain the offset of the payload
 	// TODO allow to filter by the payload
-	if len(b) >= hdrlen+dataOffset {
-		t.payload = b[hdrlen+dataOffset:]
+	if payloadOffset > 0 {
```

Comment:
can this happen? I don't know if I prefer the old check

Reply:
dataOffset is a tcp-header thing. It should be used to compute the hdrlen for tcp-headers only. I think there was a bug in the old code for sctp and udp payloads. I looked that the hdrlen was added twice, since it was included in dataOffset (but I didn't really check carefully, since I re-wrote it)

Reply:
Oh, payloadOffset must be >0 now since we return in the default: branch. I can remove the check.